pkgs/dart_mcp/lib/src/api/tools.dart

```diff
@@ -122,26 +122,344 @@ extension type Tool.fromMap(Map<String, Object?> _value) {
   /// A human-readable description of the tool.
   String? get description => _value['description'] as String?;
 
-  /// A JSON Schema object defining the expected parameters for the tool.
-  InputSchema get inputSchema => _value['inputSchema'] as InputSchema;
+  /// A JSON [ObjectSchema] object defining the expected parameters for the
+  /// tool.
+  ObjectSchema get inputSchema => _value['inputSchema'] as ObjectSchema;
 }
 
-/// A JSON Schema object defining the expected parameters for the tool.
-extension type InputSchema.fromMap(Map<String, Object?> _value) {
-  factory InputSchema({
-    Map<String, Object?>? properties,
+/// The valid types for properties in a JSON-RCP2 schema.
+enum JsonType {
+  object('object'),
+  list('array'),
+  string('string'),
+  num('number'),
+  int('integer'),
+  bool('boolean'),
+  nil('null');
+
+  const JsonType(this.typeName);
+
+  final String typeName;
+}
+
+/// A JSON Schema object defining the any kind of property.
+///
+/// See the subtypes [ObjectSchema], [ListSchema], [StringSchema],
+/// [NumberSchema], [IntegerSchema], [BooleanSchema], [NullSchema].
+///
+/// To get an instance of a subtype, you should inspect the [type] as well as
+/// check for any schema combinators ([allOf], [anyOf], [oneOf], [not]), as both
+/// may be present.
+///
+/// If a [type] is provided, it applies to all sub-schemas, and you can cast all
+/// the sub-schemas directly to the specified type from the parent schema.
+///
+/// See https://json-schema.org/understanding-json-schema/reference for the full
+/// specification.
+///
+/// **Note:** Only a subset of the json schema spec is supported by these types,
+/// if you need something more complex you can create your own
+/// `Map<String, Object?>` and cast it to [Schema] (or [ObjectSchema]) directly.
+extension type Schema.fromMap(Map<String, Object?> _value) {
+  factory Schema.withCombinators({
```

Comment:
[nit] the `withCombinators` name feels a little clunky. Can we use the unnamed constructor for this one?

Reply:
Well, it only  makes sense to use this if you are using combinators, otherwise you should construct one of the subtypes.

Reply:
I renamed this to `combined` and documented it.